trace/trace.go

```diff
@@ -37,40 +73,51 @@ type Trace struct {
 	rootSpan         *Span
 	tlfLock          sync.RWMutex
 	traceLevelFields map[string]interface{}
+	traceIDGenerator *idGenerator
 }
 
-// NewTrace creates a brand new trace. serializedHeaders is optional, and if
-// included, should be the header as written by trace.SerializeHeaders(). When
-// not starting from an upstream trace, pass the empty string here.
-func NewTrace(ctx context.Context, serializedHeaders string) (context.Context, *Trace) {
+// NewTrace creates a brand new trace. The SpanContext is optional, and if
+// included, should be the SpanContext created by an HTTPPropagator on incoming
+// requests. When not starting from an upstream teace, pass nil here.
+func NewTrace(ctx context.Context, sc *SpanContext) (context.Context, *Trace) {
 	trace := &Trace{
 		builder:          client.NewBuilder(),
 		rollupFields:     make(map[string]float64),
 		traceLevelFields: make(map[string]interface{}),
+		traceIDGenerator: newIDGenerator(),
 	}
-	if serializedHeaders != "" {
-		prop, err := propagation.UnmarshalTraceContext(serializedHeaders)
-		if err == nil {
-			trace.traceID = prop.TraceID
-			trace.parentID = prop.ParentID
-			for k, v := range prop.TraceContext {
-				trace.traceLevelFields[k] = v
-			}
-			if prop.Dataset != "" {
-				trace.builder.Dataset = prop.Dataset
-			}
+	if sc != nil {
+		trace.traceID = sc.TraceID
+		trace.parentID = sc.ParentID
+		for k, v := range sc.TraceContext {
+			trace.traceLevelFields[k] = v
+		}
+		if sc.Dataset != "" {
+			trace.builder.Dataset = sc.Dataset
 		}
 	}
 
 	if trace.traceID == "" {
-		trace.traceID = uuid.Must(uuid.NewRandom()).String()
+		var traceID [16]byte
+		trace.traceID = trace.traceIDGenerator.build(traceID[:])
 	}
 
 	rootSpan := newSpan()
 	rootSpan.isRoot = true
 	if trace.parentID != "" {
 		rootSpan.parentID = trace.parentID
 	}
+	rootSpan.spanContext = &SpanContext{
+		ParentID: rootSpan.spanID,
+		TraceID:  trace.traceID,
+	}
+
+	if sc != nil {
+		rootSpan.spanContext.Dataset = sc.Dataset
+		rootSpan.spanContext.TraceContext = sc.TraceContext
+		rootSpan.spanContext.TraceState = sc.TraceState
+	}
```

Comment:
This makes me think that there are two copies of trace context - one on the span on one in the trace. Do we need both? Spans have a link to the trace object they can use to get a trace level fields. I'll keep reading to try and understand better how the span context on a span is used